readthedocs/config/config.py

```diff
@@ -0,0 +1,487 @@
+            key=key,
+            code=code,
+            error=error_message)
+        super(InvalidConfig, self).__init__(message, code=code)
+
+
+class BuildConfig(dict):
+
+    """
+    Config that handles the build of one particular documentation.
+
+    Config keys can be accessed with a dictionary lookup::
+
+        >>> build_config['type']
+        'sphinx'
+
+    You need to call ``validate`` before the config is ready to use. Also
+    setting the ``output_base`` is required before using it for a build.
+    """
+
+    BASE_INVALID_MESSAGE = 'Invalid value for base: {base}'
+    BASE_NOT_A_DIR_MESSAGE = '"base" is not a directory: {base}'
+    NAME_REQUIRED_MESSAGE = 'Missing key "name"'
+    NAME_INVALID_MESSAGE = (
+        'Invalid name "{name}". Valid values must match {name_re}')
+    TYPE_REQUIRED_MESSAGE = 'Missing key "type"'
+    CONF_FILE_REQUIRED_MESSAGE = 'Missing key "conf_file"'
+    PYTHON_INVALID_MESSAGE = '"python" section must be a mapping.'
+    PYTHON_EXTRA_REQUIREMENTS_INVALID_MESSAGE = (
+        '"python.extra_requirements" section must be a list.')
+
+    PYTHON_SUPPORTED_VERSIONS = [2, 2.7, 3, 3.5]
+    DOCKER_SUPPORTED_VERSIONS = ['1.0', '2.0', 'latest']
+
+    def __init__(self, env_config, raw_config, source_file, source_position):
+        self.env_config = env_config
+        self.raw_config = raw_config
+        self.source_file = source_file
+        self.source_position = source_position
+        super(BuildConfig, self).__init__()
+
+    def error(self, key, message, code):
+        """Raise an error related to ``key``."""
+        source = '{file} [{pos}]'.format(
+            file=self.source_file,
+            pos=self.source_position)
+        raise InvalidConfig(
+            key=key,
+            code=code,
+            error_message='{source}: {message}'.format(source=source,
+                                                       message=message),
+            source_file=self.source_file,
+            source_position=self.source_position)
+
+    @contextmanager
+    def catch_validation_error(self, key):
+        """Catch a ``ValidationError`` and raises an ``InvalidConfig`` error."""
+        try:
+            yield
+        except ValidationError as error:
+            raise InvalidConfig(
+                key=key,
+                code=error.code,
+                error_message=str(error),
+                source_file=self.source_file,
+                source_position=self.source_position)
+
+    def get_valid_types(self):  # noqa
+        return (
+            'sphinx',
+        )
+
+    def get_valid_python_versions(self):
+        """Get all valid python versions."""
+        try:
+            return self.env_config['python']['supported_versions']
+        except (KeyError, TypeError):
+            pass
+        return self.PYTHON_SUPPORTED_VERSIONS
+
+    def get_valid_formats(self):  # noqa
+        """Get all valid documentation formats."""
+        return (
+            'htmlzip',
```

Comment:
it looks like we lost `none` in this list, do we need it?

Reply:
I moved this logic, it doesn't make much sense having this as a valid option, but in the past, we allowed `['none']` as a way of ignoring all the types, this check is now in https://github.com/rtfd/readthedocs.org/pull/4272/files#diff-2b6854693f39acfa183fca7fef27dcfbR423. BTW, we never put `none` as a valid option in the past docs.